docs/contributing/ctn-build.md

```diff
@@ -0,0 +1,33 @@
+The `Dockerfile` supports building and cross compiling docker daemon and extra
+tools using [Docker Buildx](https://github.com/docker/buildx) and [BuildKit](https://github.com/moby/buildkit).
+A [bake definition](https://github.com/docker/buildx/blob/master/docs/reference/buildx_bake.md) named `docker-bake.hcl` is in place to ease the build process:
+
+```shell
+# build binaries for the current host platform
+# output to ./bundles/binary by default
+docker buildx bake
```

Comment:
> Previously, a build would default to Linux (matching the host's architecture, where "host" being "where the daemon runs").

This has not changed, currently `make [binary|dynbinary|cross]` use Docker to build: https://github.com/moby/moby/blob/7624f8aeb1836cd75a0912d9614930e8a59eaf4e/Makefile#L171-L179

We are just using bake as high level build command to simplify the workflow.

> While I can see value in having `buildx bake` as a replacement for `make` to build binaries for the _current_ platform (but in a sandboxed environment), for this project in particular, that's a bit of a weird assumption, because effectively we only support Linux and Windows (and only on a limited set of architectures)

Windows users can't use `make` (except on WSL) but buildx binary is available on Windows so it's a benefit to rely on it I think.

> This default changes the default to (in many cases) cross-compile, and, e.g. on a macOS machine, produces no useful output at all

I'm not sure I understand, you mean `docker buildx bake` will not produce anything on MacOS? Or you mean `docker buildx bake --set *.platform=local` should output something?

Or you mean that building on M1 machine will cross-comp? Because it will not as it will pull the linux/arm64 image and build natively on it.

Can you explain a bit more?

> if I would run it on a "native" platform, then the default would be a static binary (mostly useful for testing) as this repository does not have recipes to build for specific distros (which is all part of the packaging repositories).

This is the current behavior when you type `make` currently:

```
$ make
docker build  --build-arg=GO_VERSION  -f "Dockerfile" --output=bundles/ --target=binary --build-arg VERSION --build-arg DOCKER_GITCOMMIT --build-arg PRODUCT --build-arg PLATFORM --build-arg DEFAULT_PRODUCT_LICENSE --build-arg PACKAGER_NAME .
```

> Overall, I think the _main purpose_ of the binaries, is;
> 
>     * to be used as a dependency for `make dev` / `make shell`; to have a binary available within the "dev" container (`make BIND_DIR=. binary shell`, or within the container, `hack/make.sh binary && make install`)
> 
>     * for CI (running docker-in-docker)
> 
>     * for Windows (doesn't support docker-in-docker); to build the binary and run it as part of CI

Yes the main purpose is to have a better cross compilation support because today the Dockerfile does not support `--platform` with BuildKit to build efficiently against any arch and so we need to rely on obscure env vars `DOCKER_CROSSPLATFORMS`, `CROSS` instead of automatic platform ARGs in the global scope like `TARGETPLATFORM` and so current `cross` stage will build against all platforms which increase build time.

Same for cross pkgs hardcoded so you have to modify the Dockerfile to add support for extra archs: https://github.com/moby/moby/blob/7624f8aeb1836cd75a0912d9614930e8a59eaf4e/Dockerfile#L108-L112 https://github.com/moby/moby/blob/7624f8aeb1836cd75a0912d9614930e8a59eaf4e/Dockerfile#L137-L147

With these changes it's mostly transparent and will help on docker-ce-packaging: https://github.com/docker/docker-ce-packaging/pull/718/commits/dacaeb69fd404bcb3c3215c31bd2fff78b26f2c7 as well as https://github.com/rumpl/moby/pull/21#issuecomment-1189818960

> Currently, this includes "build containerd and runc", which basically shouldn't be part of that (containerd, runc are separate projects, and should be packaged separately like we do for the .deb and .rpm)

Yes the `cross` job in `ci` workflow currently build binaries and extra tools for testing purpose and to make sure nothing is broken. But agree that in a follow-up we should remove containerd, runc, ... and just grab them from their respective location. Same applies on docker-ce-packaging.

> For _non-static_ binaries (`.deb`, `.rpm`) things are somewhat more complicated;
> 
>     * traditional packaging tools for deb and rpm expect a source tarball as input (no dependency on git repo), and building to be done as part of that tooling
> 
>     * they should be built in an environment that matches the target distro (due to dynamic linking, and dependencies (or their names) to differ between distros)

Nothing change you can build dynamic binaries as before in your spec files like in docker-ce-packaging: https://github.com/docker/docker-ce-packaging/pull/718/commits/150d57b7e7d1e61ddac91bea60f5b3fddbff3c78

> 
>     * our pipeline builds those _within_ a docker container (but does not use docker as part of the "build" by the deb/rpm tools)

Yes as it is today. Do you mean you think about integrating packaging of `.deb` and `.rpm` on moby repo directly?

Reply:
> as well as [rumpl#21 (comment)](https://github.com/rumpl/moby/pull/21#issuecomment-1189818960)

@thaJeztah I updated https://github.com/rumpl/moby/pull/21 to use changes of this PR in https://github.com/rumpl/moby/pull/24 if you're interested to take a look.